lib/libc/gen/auth_compat.3
.\" Copyright (c) 1997 Berkeley Software Design, Inc. All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by Berkeley Software Design,
.\"	Inc.
.\" 4. The name of Berkeley Software Design, Inc.  may not be used to endorse
.\"    or promote products derived from this software without specific prior
.\"    written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY BERKELEY SOFTWARE DESIGN, INC. ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL BERKELEY SOFTWARE DESIGN, INC. BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	BSDI auth_compat.3,v 2.3 1998/01/28 02:49:23 prb Exp
.Dd "March 20, 1997"
.Dt AUTH_COMPAT 3
.Os 
.Sh NAME
.Nm auth_approve ,
.Nm auth_check ,
.Nm auth_clropts ,
.Nm auth_env ,
.Nm auth_response ,
.Nm auth_rmfiles ,
.Nm auth_scan ,
.Nm auth_script ,
.Nm auth_script_data ,
.Nm auth_setopt ,
.Nm auth_value
.Nd deprecated BSD Authentication interface
.Sh SYNOPSIS
.Fd #include <login_cap.h>

.Ft int
.Fn auth_approve "login_cap_t *lc" "char *fullname" "char *service"
.Ft int
.Fn auth_check "char *fullname" "char *class" "char *style" "char *service" "int *stat"
.Ft void
.Fn auth_clropts
.Ft void
.Fn auth_env
.Ft int
.Fn auth_response "char *fullname" "char *class" "char *style" "char *service" "int *stat" "char *challenge" "char *response"
.Ft void
.Fn auth_rmfiles
.Ft int
.Fn auth_scan "int okay"
.Ft int
.Fn auth_script "char *path" "..."
.Ft int
.Fn auth_script_data "char *data" "int nbytes" "char *path" "..."
.Ft int
.Fn auth_setopt "char *name" "char *value"
.Ft char *
.Fn auth_value "char *name"
.Sh DESCRIPTION
These functions provide the deprecated interface to the BSD Authentication
system.  They are only provided for backwards compatibility with existing
software and should not be used in new code.
.Sh
BSD Authentication is performed through external login scripts with access
controlled by the
.Pa /etc/login.conf
file.  See
.Xr login.conf 5
for a description of the interface to these external programs.
.Sh THE PARAMETERS
The following parameters are used in the various routines:
.Bl -tag -width fullnamexx
.It Ar challenge
A null terminated string representing the challenge issued to the user.
.It Ar class
The users login class.  See
.Xr login.conf 5
for more details.
.It Ar data
A pointer to arbitrary data to be passed to the login script on
file descriptor 3 (the
.Dq backchannel ) .
The data may contain NULs.
.It Ar fullname
The full name of the user being authenticated, including any instance.
.It Ar lc
A pointer to the
.Ar login_cap_t
returned by the
.Xr login_getclass 3
routine.  This value can be NULL in which case the
.Ar fullname
parameter is used to lookup the login class for the user.
.It Ar name
The name of a variable or option.
.It Ar nbytes
The number of bytes the
.Ar data
parameter is pointing to.
.It Ar okay
The value to return if no explicit authentication or rejection
message is received from the login_script.
.It Ar path
.It Ar response
A null terminated string representing the response to the challenge that
was issued to the user.
.It Ar service
The type of service being requested of the login script.
See
.Xr login.conf 5
for a description of the various services.
A NULL may be passed to indicate the default service should be used.
The default service requires that standard input, output, and error
be available for their normal interaction with the user.
.It Ar stat
If not NULL, the full return status is stored to the location to which
.Ar stat points.
The return values are a subset of the status such that any return value
greater than 0 indicates the authentication was successful.  The
additional information contained in the full status indicates how to
treat a failure.
.It Ar style
The style of authentication to be used.
See
.Xr login.conf 5
for a description of the various styles.
This value must not be NULL, there is no hard coded default style.
.It Ar value
The value to be associated with the specified name.
.El
.Sh STATUS
The status is made up of one or more the following values ORed together:
.Bl -tag -width fullnamexx
.It Li AUTH_OKAY
The user was authenticated.
.It Li AUTH_ROOTOKAY
The user was authenticated with a root instance.
.It Li AUTH_SECURE
The user was authenticated via a mechanism which is not subject to
eavesdropping attacks (such as provided by token cards).
.It Li AUTH_SILENT
Do not report an error, the user was not authenticated for access and
was not expected to be.  This is returned by login scripts that allow
changing of the users password, for instance.
This value is stripped off for normal returns.
.It Li AUTH_CHALLENGE
The user was not authenticated for access and a challenge was issued.
The challenge should be displayed to the user, a response retrieved,
and the result verified.
This value is stripped off for normal returns.
.El
.Sh THE FUNCTIONS
.Bl -tag -width fullnamexx
.It Fn auth_approve
Run the approval script for the specified service, a return greater than 0
indicates the user was approved.
.It Fn auth_check
Run the login script specified by the
.Ar style
for the user.
A return value greater than 0 indicates success.  Use the value
pointed to by
.Ar stat
to determine if a rejection was silent or a challenge was issued.
Use the
.Fn auth_value
function to retrieve any generated challenge.
.It Fn auth_clropts
Clear any options stored for sending to the login script.
.It Fn auth_env
Set any environment variables specified by the last login_script to
have run.
.It Fn auth_response
Verify a response from the user for the specified challenge.  The challenge
can be the NULL string if the user provided a response prior to the
issuing of challenge (such as from a PortMaster).
.It Fn auth_rmfiles
Remove any files generated by the login script.  This should only be
done if the user was ultimately denied access.
.It Fn auth_scan
Scan the data returned from a login script and return the status determined.
If no status was set, return
.Ar okay .
(Note that
.Ar okay
should always be 0 when checking authentication, approval scripts
use the value of
.Li AUTH_OKAY
since no return from them implies the user was accepted.)
.It Fn auth_script
Execute the requested script, reading responses on file descriptor 3.
Any options specified by the
.Fn auth_setopt
function are inserted between the 0th and first argument.
Each option is preceeded by the
.Fl v
flag.
.It Fn auth_script_data
The same as
.Fn auth_script
except the specified data is written to file descriptor 3 prior to
reading the response.  The data typically should be of the form
.Dq challenge\0response\0 .
.It Fn auth_setopt
When calling a login script, insert the following option on the command line:
.Li Fl v Ar name=value
.It Fn
Search the response from the last login script for the value associated
with the given name.  See
.Xr login.conf 5
for a list of possible names.
.El 
.Sh ISSUES
These routines use global state so only one BSD Authentication session
can be effectively used at one time.  While
.Fn auth_check
provides a simple single function call authentication
interface, it depends on standard input, output and error.
The programmatic interface, in particular when challenges are issued,
is complex and overly burdensome.  The new BSD Authentication interface
provides a much better interface.
.Sh SEE ALSO
.Xr authenticate 3 ,
.Xr auth_subr 3 ,
.Xr login_cap 3 ,
.Xr login.conf 5


lib/libc/gen/auth_compat.c
/*-
 * Copyright (c) 1995,1997 Berkeley Software Design, Inc. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by Berkeley Software Design,
 *	Inc.
 * 4. The name of Berkeley Software Design, Inc.  may not be used to endorse
 *    or promote products derived from this software without specific prior
 *    written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY BERKELEY SOFTWARE DESIGN, INC. ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL BERKELEY SOFTWARE DESIGN, INC. BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 *	BSDI auth_compat.c,v 2.3 1999/09/08 04:10:40 prb Exp
 */
#include <sys/param.h>
#include <sys/time.h>
#include <sys/resource.h>
#include <sys/socket.h>
#include <sys/wait.h>

#include <ctype.h>
#include <err.h>
#include <fcntl.h>
#include <paths.h>
#include <pwd.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <syslog.h>
#include <unistd.h>

#include <login_cap.h>

struct rmfiles {
	struct rmfiles	*next;
	char		*file;
};

struct authopts {
	struct authopts	*next;
	char		*opt;
};

#define	MAXSPOOLSIZE	(8*1024)	/* Spool up to 8K of back info */

static char spool[MAXSPOOLSIZE];
static int spoolindex = 0;
static struct rmfiles *rmroot = NULL;
static struct authopts *optroot = NULL;

static void add_rmlist(char *);
static void spooldata(int);
static int _auth_script(char *, int, char *, va_list);

/*
 * Set up a known environment for all authentication scripts.
 */
static char *auth_environ[] = {
	"PATH=" _PATH_DEFPATH,
	"SHELL=" _PATH_BSHELL,
	NULL,
};

int
auth_check(char *fullname, char *class, char *style, char *service, int *stat)
{
	char path[MAXPATHLEN];
	int _stat;

	if (!stat)
		stat = &_stat;

	*stat = 0;

	if (!style)
		return (-1);
	if (!service)
		service = LOGIN_DEFSERVICE;

	snprintf(path, sizeof(path), _PATH_AUTHPROG "%s", style);
	if (auth_script(path, style, "-s", service, fullname, class, 0))
		return (*stat = 0);
	return ((*stat = auth_scan(0)) & AUTH_ALLOW);
}

int
auth_response(char *fullname, char *class, char *style, char *service,
    int *stat, char *challenge, char *response)
{
	char *data;
	char path[MAXPATHLEN];
	int _stat, datalen;

	if (!stat)
		stat = &_stat;

	*stat = 0;

	if (!style)
		return (-1);
	if (!service)
		service = LOGIN_DEFSERVICE;

	datalen = strlen(challenge) + strlen(response) + 2;

	if ((data = malloc(datalen)) == NULL) {
		syslog(LOG_ERR, "%m");
		warnx("internal resource failure");
		return (-1);
	}
	snprintf(data, datalen, "%s%c%s", challenge, 0, response);

	snprintf(path, sizeof(path), _PATH_AUTHPROG "%s", style);
	if (auth_script_data(data, datalen, path, style, "-s", service,
	    fullname, class, 0))
		*stat = 0;
	else
		*stat = auth_scan(0);
	free(data);

	return (*stat & AUTH_ALLOW);
}

void
auth_env()
{
	char *line;
	char *name;

    	for (line = spool; line < spool + spoolindex;) {
		if (!strncasecmp(line, BI_SETENV, sizeof(BI_SETENV)-1)) {
			line += sizeof(BI_SETENV) - 1;
			if (isblank(*line)) {
				for (name = line; isblank(*name); ++name)
					;
				for (line = name; *line && !isblank(*line);
				    ++line)
					;
				if (*line)
					*line++ = '\0';
				for (; isblank(*line); ++line)
					;
				if (*line != '\0' && setenv(name, line, 1))
					warn("setenv(%s, %s)", name, line);
			}
		} else
		if (!strncasecmp(line, BI_UNSETENV, sizeof(BI_UNSETENV)-1)) {
			line += sizeof(BI_UNSETENV) - 1;
			if (isblank(*line)) {
				for (name = line; isblank(*name); ++name)
					;
				for (line = name; *line && !isblank(*line);
				    ++line)
					;
				if (*line)
					*line++ = '\0';
				unsetenv(name);
			}
		}
		while (*line++)
			;
	}
}

char *
auth_value(char *what)
{
	char *line;
	char *name;
	char *value;
	int n;

    	for (line = spool; line < spool + spoolindex;) {
		if (!strncasecmp(line, BI_VALUE, sizeof(BI_VALUE)-1)) {
			line += sizeof(BI_VALUE) - 1;
			if (isblank(*line)) {
				for (name = line; isblank(*name); ++name)
					;
				for (line = name; !isblank(*line); ++line)
					;
				if (*line)
					*line++ = '\0';
				if (strcmp(name, what))
					continue;
				for (; isblank(*line); ++line)
					;
				value = strdup(line);
				if (value == NULL)
					return (NULL);

				/*
				 * XXX - There should be a more standardized
				 * routine for doing this sort of thing.
				 */
				for (line = name = value; *line; ++line) {
					if (*line == '\\') {
						switch (*++line) {
						case 'r':
							*name++ = '\r';
							break;
						case 'n':
							*name++ = '\n';
							break;
						case 't':
							*name++ = '\t';
							break;
						case '0': case '1': case '2':
						case '3': case '4': case '5':
						case '6': case '7':
							n = *line - '0';
							if (isdigit(line[1])) {
								++line;
								n <<= 3;
								n |= *line-'0';
							}
							if (isdigit(line[1])) {
								++line;
								n <<= 3;
								n |= *line-'0';
							}
							break;
						default:
							*name++ = *line;
							break;
						}
					} else
						*name++ = *line;
				}
				*name = '\0';
				return (value);
			}
		}
		while (*line++)
			;
	}
	return (NULL);
}

void
auth_rmfiles()
{
	while (rmroot) {
		unlink(rmroot->file);
		rmroot = rmroot->next;
	}
}

int
auth_script(char *path, ...)
{
	va_list ap;
	int r;

	va_start(ap, path);

	r = _auth_script(NULL, 0, path, ap);

	va_end(ap);

	return (r);
}

int
auth_script_data(char *data, int nbytes, char *path, ...)
{
	va_list ap;
	int r;

	va_start(ap, path);

	r = _auth_script(data, nbytes, path, ap);

	va_end(ap);

	return (r);
}

static int
_auth_script(char *data, int nbytes, char *path, va_list ap)
{
	struct authopts *e;
	pid_t pid;
	int status;
	int pfd[2];
	int argc;
	char *argv[64];		/* 64 args should more than enough */
#define	Nargc	(sizeof(argv)/sizeof(argv[0]))

	argc = 0;
	if ((argv[argc] = va_arg(ap, char *)) != NULL)
		++argc;

	for (e = optroot; e != NULL; e = e->next) {
		if (argc < Nargc - 2) {
			argv[argc++] = "-v";
			argv[argc++] = e->opt;
		} else {
			syslog(LOG_ERR, "too many authentication options");
			return (-1);
		}
	}
	while (argc < Nargc - 1 && (argv[argc] = va_arg(ap, char *)))
		++argc;

	if (argc >= Nargc - 1 && va_arg(ap, char *)) {
		syslog(LOG_ERR, "too many arguments");
		return (-1);
	}

	argv[argc] = NULL;

	spoolindex = 0;

	if (secure_path(path) < 0) {
		syslog(LOG_ERR, "%s: path not secure", path);
		warnx("invalid script: %s", path);
		return (-1);
	}

	if (socketpair(PF_LOCAL, SOCK_STREAM, 0, pfd) < 0) {
		syslog(LOG_ERR, "unable to create backchannel %m");
		warnx("internal resource failure");
		return (-1);
	}

	switch (pid = fork()) {
	case -1:
		close(pfd[0]);
		close(pfd[1]);
		syslog(LOG_ERR, "%s: %m", path);
		warnx("internal resource failure");
		return (-1);
	case 0:
#define	COMM_FD	3
		close(pfd[0]);
		if (pfd[1] != COMM_FD) {
			if (dup2(pfd[1], COMM_FD) < 0)
				err(1, "dup of backchannel");
			close(pfd[1]);
		}

		for (status = getdtablesize() - 1; status > COMM_FD; status--)
			close(status);

		execve(path, argv, auth_environ);
		syslog(LOG_ERR, "%s: %m", path);
		err(1, path);
	default:
		close(pfd[1]);
		if (data && nbytes)
			write(pfd[0], data, nbytes);
		spooldata(pfd[0]);
		close(pfd[0]);
		if (waitpid(pid, &status, 0) < 0) {
			syslog(LOG_ERR, "%s: waitpid: %m", path);
			warnx("internal failure");
			return (-1);
		}

		if (!WIFEXITED(status) || WEXITSTATUS(status))
			return (-1);
	}
	return (0);
}

int
auth_scan(int okay)
{
	char *line;

    	for (line = spool; line < spool + spoolindex;) {
		if (!strncasecmp(line, BI_REJECT, sizeof(BI_REJECT)-1)) {
			line += sizeof(BI_REJECT) - 1;
			if (!*line || *line == ' ' || *line == '\t') {
				while (*line == ' ' || *line == '\t')
					++line;
				if (!strcasecmp(line, "silent"))
					return (AUTH_SILENT);
				if (!strcasecmp(line, "challenge"))
					return (AUTH_CHALLENGE);
			}
			return (0);
		} else if (!strncasecmp(line, BI_AUTH, sizeof(BI_AUTH)-1)) {
			line += sizeof(BI_AUTH) - 1;
			if (!*line || *line == ' ' || *line == '\t') {
				while (*line == ' ' || *line == '\t')
					++line;
				if (*line == '\0')
					okay |= AUTH_OKAY;
				else if (!strcasecmp(line, "root"))
					okay |= AUTH_ROOTOKAY;
				else if (!strcasecmp(line, "secure"))
					okay |= AUTH_SECURE;
			}
		}
		else if (!strncasecmp(line, BI_REMOVE, sizeof(BI_REMOVE)-1)) {
			line += sizeof(BI_REMOVE) - 1;
			while (*line == ' ' || *line == '\t')
				++line;
			if (*line)
				add_rmlist(line);
		}
		while (*line++)
			;
	}

	return (okay);
}

int
auth_setopt(char *n, char *v)
{
	struct authopts *e;

	if ((e = malloc(sizeof(*e))) == NULL)
		return (-1);
	if ((e->opt = malloc(strlen(n)+strlen(v)+2)) == NULL) {
		free(e);
		return (-1);
	}
	sprintf(e->opt, "%s=%s", n, v);
	e->next = optroot;
	optroot = e;
	return(0);
}

void
auth_clropts()
{
	struct authopts *e;
	while ((e = optroot) != NULL) {
		optroot = optroot->next;
		free(e->opt);
		free(e);
	}
}

static void
spooldata(int fd)
{
	int r;
	char *b;

	while (spoolindex < sizeof(spool) - 1) {
		r = read(fd, spool + spoolindex, sizeof(spool)-spoolindex);
		if (r <= 0) {
			spool[spoolindex] = '\0';
			return;
		}
		b = spool + spoolindex;
		spoolindex += r;
		/*
		 * Go ahead and convert newlines into NULs to allow
		 * easy scanning of the file.
		 */
		while(r-- > 0)
			if (*b++ == '\n')
				b[-1] = '\0';
	}

	syslog(LOG_ERR, "Overflowed backchannel spool buffer");
	err(1, "System error in authentication program");
}

static void
add_rmlist(char *file)
{
	struct rmfiles *rm;

	if ((rm = malloc(sizeof(struct rmfiles))) == NULL) {
		syslog(LOG_ERR, "Failed to allocate rmfiles: %m");
		return;
	}
	if (!(file = strdup(file))) {
		syslog(LOG_ERR, "Failed to duplicate rmfile: %m");
		return;
	}
	rm->file = file;
	rm->next = rmroot;
	rmroot = rm;
}
